cudnn*_cudatoolkit*: use public NVIDIA mirror

lukeadams opened the pull request:
###### Motivation for this change
This PR dramatically simplifies the use of cudnn libraries on NixOS, which is used by most machine learning packages for gpu acceleration.

Previously, you would have to go through the NVIDIA developer portal and manually use nix-prefetch to add the tarballs to the store. 
- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Commit:
cudnn*_cudatoolkit*: use NVIDIA mirror instead of requireFile

jackpot1136 commented:
please push thought

grahamc commented:
It looks like we _are_ allowed to distribute this lib as long as we include this copyright notice: https://gitlab.com/nvidia/cuda/blob/centos7/LICENSE#L1

lukeadams commented:
Would including the license in the meta be acceptable? Or print it while the package is being built?

lukeadams commented on pkgs/development/libraries/science/math/cudnn/default.nix:
```diff
@@ -9,9 +9,10 @@ in
     cudatoolkit = cudatoolkit7;
-    srcName = "cudnn-${cudatoolkit.majorVersion}-linux-x64-v${version}-prod.tgz";
+    srcName = "cudnn-${cudatoolkit.majorVersion}-linux-x64-v4.0-prod.tgz";
     sha256 = "0zgr6qdbc29qw6sikhrh6diwwz7150rqc8a49f2qf37j2rvyyr2f";
```
That's weird – I tested all of them and the hashes matched. 
Maybe the url was corrected to match the format of the newer ones. I'll look into it this weekend.

Commit:
cudnn_cudatoolkit7: update hash since public mirror is different

lukeadams commented:
@grahamc I fixed the hash issue and re-tested, all successfully downloaded and built. Sorry about that.

I added BSD3 to the license, but since the blobs being downloaded are unfree I left that in as well.

lukeadams commented:
Rebased onto master. manually building affected packages with nix-build works (Changes not top level so nox-review doesn't pick it up)

Still not sure what the license should be...

lukeadams commented:
Rebased as of today

orivej commented on pkgs/development/libraries/science/math/cudnn/generic.nix:
```diff
@@ -49,7 +44,7 @@ stdenv.mkDerivation rec {
     description = "NVIDIA CUDA Deep Neural Network library (cuDNN)";
     homepage = "https://developer.nvidia.com/cudnn";
-    license = licenses.unfree;
+    license = [ licenses.unfree licenses.bsd3 ];
```
Does BSD3 apply to anything brought by this derivation, or is it just the license of Dockerfiles where you have found source URLs?

lukeadams replied:
That's just the license in the Dockerfiles repo, which I now realize is _definitely_ not the cudnn license. [It looks like this is the actual license on the NVIDIA site.](http://docs.nvidia.com/deeplearning/sdk/cudnn-sla/index.html)
I can add the license to licenses.nix, something like
```
nvidia-cudnn = {
  fullName  = "NVIDIA SOFTWARE LICENSE AGREEMENT for cuDNN";
  url       = "http://docs.nvidia.com/deeplearning/sdk/pdf/cuDNN-SLA.pdf";
  free      = false;
};
```

[`cudnn` and friends have been posted to AUR with license proprietary](https://aur.archlinux.org/packages/cudnn6/)

orivej replied:
IMHO just keeping the old `unfree` is fine. Then it looks ready to merge.

lukeadams commented:
@orivej I dropped the bsd3 commit

lukeadams commented:
@orivej Rebased 2018-02-10

lukeadams commented:
Rebased 2018-03-01

lukeadams commented:
Rebased 2018-03-14, fixed merge conflict. Ready for review/merge

lukeadams commented:
cc @orivej @grahamc 
Rebased 2018-05-13

Since nox-review doesn't pick up the changes, the packages were tested with `nix-build '<nixpkgs>' -I nixpkgs=/media/manual/nixpkgs-checkout/nixpkgs --no-out-link -A cudnn_cudatoolkit7 -A cudnn_cudatoolkit75 -A cudnn6_cudatoolkit8 -A cudnn_cudatoolkit8 -A cudnn_cudatoolkit9`, which completes successfully.

lukeadams commented:
@orivej Are there any remaining issues? I just rebased and resolved the merge conflicts so it's still good to go afaik.

jb55 reviewed (approved):
👍

lukeadams commented:
cc @matthewbauer I just rebased it again. Any chance we could get this merged?

lukeadams commented:
@matthewbauer orivej and I were talking about adding the CUDNN license on the nvidia site, [but decided against it in favor of the license on the Dockerfile repo the source url is from.
](https://github.com/NixOS/nixpkgs/pull/31551#discussion_r162788062)
[Some debian maintainers were unsure of the redist status](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=862524), so I'll probably leave it as unfree. 
Would there be a way to have Hydra build+cache `tensorflowWithCuda` with it set as such?

orivej commented:
Thank you and sorry for the delay!

lukeadams commented:
@matthewbauer Hey sorry to ping you again, but is there a way to have Hydra build the cudnn/CUDA/tensorflowWithGpu packages but not cache the output (i.e. so it gets listed on hydra.nixos.org, but without allowing redistribution)?

matthewbauer commented:
No, I don't think there is a way to do that. Nix doesn't really separate out licenses of "buildPackages" vs "hostPackages". Everything either evaluates or not. This ties into some of the cross stuff & what I think @nh2 is working on. At one point in time you could put ```unfreeRedistributable``` in the license and Hydra would evaluate it - but that appears to have been removed. If you had your own Hydra instance you could still do something like that with:

```
let lib = import <nixpkgs/lib>;

in {
  allowUnfree = false;
  whitelistedLicenses = [ lib.licenses.unfreeRedistributable ];
}
```

Or make up some other license for it.
